docs/proposals/volume-selectors.md

```diff
@@ -0,0 +1,295 @@
+distinguish between instances of these types.  An administrator might decide
+to think of these volumes as follows:
+
+1.  Cold HDD - `tin`
+2.  Throughput optimized HDD - `bronze`
+3.  General purpose SSD - `silver`
+4.  Provisioned IOPS SSD - `gold`
+
+This is not the only dimension that EBS volumes can differ in.  Let's simplify
+things and imagine that AWS has two availability zones, `east` and `west`. Our
+administrators want to differentiate between volumes of the same type in these
+two zones, so they create a taxonomy of volumes like so:
+
+1.  `tin-west`
+2.  `tin-east`
+3.  `bronze-west`
+4.  `bronze-east`
+5.  `silver-west`
+6.  `silver-east`
+7.  `gold-west`
+8.  `gold-east`
+
+Another administrator of the same cluster might label things differently,
+choosing to focus on the business role of volumes.  Say that the data
+warehouse department is the sole consumer of the cold HDD type, and the DB as
+a service offering is the sole consumer of provisioned IOPS volumes.  The
+administrator might decide on the following taxonomy of volumes:
+
+1.  `warehouse-east`
+2.  `warehouse-west`
+3.  `dbaas-east`
+4.  `dbaas-west`
+
+There are any number of ways an administrator may choose to distinguish
+between volumes.  Labels are used in Kubernetes to express the user-defined
+properties of API objects and are a good fit to express this information for
+volumes.  In the examples above, administrators might differentiate between
+the classes of volumes using the labels `business-unit`, `volume-type`, or
+`region`.
+
+Label selectors are used through the Kubernetes API to describe relationships
+between API objects using flexible, user-defined criteria.  It makes sense to
+use the same mechanism with persistent volumes and storage claims to provide
+the same functionality for these API objects.
+
+## Proposed Design
+
+We propose that:
+
+1.  A new field called `LabelSelector` be added to the
+    `PersistentVolumeClaimSpec` type
+2.  The persistent volume controller be modified to account for this selector
+    when determining the volume to bind to a claim
+
+### Persistent Volume Selector
+
+Label selectors are used throughout the API to allow users to express
+relationships in a flexible manner.  The problem of selecting a volume to
+match a claim fits perfectly within this metaphor.  Adding a label selector to
+`PersistentVolumeClaimSpec` will allow users to label their volumes with
+criteria important to them and select volumes based on these criteria.
+
+```go
+// PersistentVolumeClaimSpec describes the common attributes of storage devices
+// and allows a Source for provider-specific attributes
+type PersistentVolumeClaimSpec struct {
+    // Contains the types of access modes required
+    AccessModes []PersistentVolumeAccessMode `json:"accessModes,omitempty"`
+    // PersistentVolumeSelector is a selector which must be true for the claim to bind to a volume
+    LabelSelector *unversioned.LabelSelector `json:"labelSelector,omitempty"`
```

Comment:
There seems to be some inconsistency between `LabelSelector` and `PersistentVolumeSelector` throughout.